### PR TITLE
cleanup: document RetryPolicy base class

### PR DESCRIPTION
**Pull request by coryan**
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/4958)

### CODE REVIEW

File: google/cloud/internal/retry_policy.h [hunk 1]
```diff
@@ -26,13 +26,18 @@ inline namespace GOOGLE_CLOUD_CPP_NS {
 namespace internal {
 /**
  * Define the interface for retry policies.
- *
  */
 class RetryPolicy {
  public:
   virtual ~RetryPolicy() = default;
+
+  /// Handle a failure, return `true` if the retry loop can continue.
```
Comment (devbww) [inline google/cloud/internal/retry_policy.h:34]: `s/, return/. Return./`

File: google/cloud/internal/retry_policy.h [hunk 2]
```diff
@@ -26,13 +26,18 @@ inline namespace GOOGLE_CLOUD_CPP_NS {
...
   virtual bool OnFailure(Status const&) = 0;
+
+  /// Return `true` if no more retries should be attempted.
   virtual bool IsExhausted() const = 0;
+
+  /// Return `true` if the error is not retryable.
```
Comment (devbww) [inline google/cloud/internal/retry_policy.h:40]: While any documentation on an interface is probably better than none, I find it difficult to distinguish "_return `true` if the retry loop can continue_" from "_return `true` if no more retries should be attempted_" and "_return `true` if the error is not retryable._"  They all sound like the same thing.

Comment (coryan) [reply google/cloud/internal/retry_policy.h:40]: Yes, that is a consequence of how these are intended to be used, and are probably code smells.... basically the loop is like this (paraphrasing):

```
while (!policy->IsExhausted()) {
  auto status = try_rpc();
  if (status.ok()) return status;
  if (!OnFailure(status)) return ModifyError(status);
  // sleep, which may exhaust the policy, even if it was not exhausted in the last call. 
}
return SomeError();
}
```

Happy to improve the descriptions if you have suggestions.

Comment (devbww) [reply google/cloud/internal/retry_policy.h:40]: Perhaps, then, the documentation might be better with the three functions grouped under a single comment, where that comment shows how they will be called in the retry loop?  That might be the easiest way to distinguish their roles.

The example loop you gave looks good, for example, although the `OnFailure(status)` is missing its `policy->`, and `IsPermanentFailure()` isn't mentioned.

File: google/cloud/internal/retry_policy.h (hunk 1 above)
Comment (coryan) [reply google/cloud/internal/retry_policy.h]: Removed.

File: google/cloud/internal/retry_policy.h (hunk 2 above)
Comment (coryan) [reply google/cloud/internal/retry_policy.h]: PTAL

Comment (devbww) [reply google/cloud/internal/retry_policy.h]: Much clearer, I reckon.  Thanks.

**Pull request merged**